provider/resources.go

```diff
@@ -7122,3 +7113,89 @@ func Provider() *tfbridge.ProviderInfo {
 
 	return &prov
 }
+
+// Apply provider tags to an individual resource.
+//
+// Historically, Pulumi has struggles to handle the "tags" and "tags_all" fields correctly:
+// - https://github.com/pulumi/pulumi-aws/issues/2633
+// - https://github.com/pulumi/pulumi-aws/issues/1655
+//
+// terraform-provider-aws has also struggled with implementing their desired behavior:
+// - https://github.com/hashicorp/terraform-provider-aws/issues/29747
+// - https://github.com/hashicorp/terraform-provider-aws/issues/29842
+// - https://github.com/hashicorp/terraform-provider-aws/issues/24449
+//
+// The Terraform lifecycle simply does not have a good way to map provider configuration
+// onto resource values, so terraform-provider-aws is forced to work around limitations in
+// unreliable ways. For example, terraform-provider-aws does not apply tags correctly with
+// -refresh=false.
+//
+// This gives pulumi the same limitations by default. However, unlike Terraform, Pulumi
+// does have a clear way to insert provider configuration into resource properties:
+// Check. By writing a custom check function that applies "default_tags" to "tags" before
+// the Terraform provider sees any resource configuration, we can give a consistent,
+// reliable and good experience for Pulumi users.
+func applyTags(
+	ctx context.Context, config resource.PropertyMap, meta resource.PropertyMap,
+) (resource.PropertyMap, error) {
+	var defaultTags awsShim.TagConfig
+
+	unknown := func() (resource.PropertyMap, error) {
+		current := config["tags"]
+		if current.IsOutput() {
+			output := current.OutputValue()
+			output.Known = false
+			config["tags"] = resource.NewOutputProperty(output)
```

Comment:
This mutates config besides returning it..

Reply:
It does. `applyTags` is only called like this:

```go
	if check := res.Schema.PreCheckCallback; check != nil {
		news, err = check(ctx, news, p.configValues.Copy())
		if err != nil {
			return nil, err
		}
	}
```

[definition](https://github.com/pulumi/pulumi-terraform-bridge/blob/86acdc26fa8cd7bb6a081162de0771ee715554e8/pkg/tfbridge/provider.go#L570-L576)

It's ok to mutate. Its an "in/out" parameter in practice, but it's not idiomatic go to write `func(ctx context.Context, config *resource.PropertyMap, meta resource.PropertyMap) error`.